### DISCUSSION TITLE
DependencyValues thread safe?

### DISCUSSION THREAD

**Discussion by DavidKmn** [pointfreeco/swift-composable-architecture#1182] [Q&A]
Was looking into the recent developments on DI you were doing on the `concurrency-updates-proto` branch. Was wondering is there a reason there is no locking in place for accessing the dependencies stored in storage? Seems to me that this can cause problems if multiple Tasks start accessing DependencyValues from different threads.

**Comment by stephencelis** [comment]
@DavidKmn We believe it should be, due to the fact that it's a value type that is only mutated via an internal `TaskLocal`, and the general practice that all dependency values should be sendable.

If anyone find issues with the current approach, though, we'd love to patch them up!

**Reply by DavidKmn** [reply]
I'm thinking of using something similar in a non TCA project, with non-spendable dependencies and without the task local modifier for mutations since this would not fit too well in an MVVM approach. Would you suggest having something like a NSRecursiveLock to protect access to deps in this case?

**Reply by DavidKmn** [reply]
Maybe I am misunderstanding what you mean @stephencelis. So will be more concrete. Let's say we have a `NetworkClient` that is `Sendable` and is modelled as a `struct` with closures annotated with `@Sendable`. Now my question is about the access to this client. On first access is the `NetworkClient` is uninitialised and we are running in a non test target we would need to access the live version. `DependencyValues` then returns either: 
 1) liveValues known at compile time and accessed through opening an existential (the else case in the getter)
 2) Or values found in the `storage` that is marked as `@TaskLocal`, from my understanding used mostly for injection in testing?
 
So are you saying 1) is safe for concurrent access since its just a function to access a an underlying value, and 2) is safe since it only allows for mutation via `@TaskLocal`?

**Reply by stephencelis** [reply]
@DavidKmn In our specific case we've cordoned off all mutation to copies of the value, done at the task-local level, so we think any mutation is free of race conditions (and we also think that in typical usage, this mutation will always done on the same thread, though it doesn't _need_ to be). We think concurrent reads are OK because we're dealing with a value type that's effectively a static let. Your mileage may vary depending on the API you expose and how you interact with it.

Of course we may also be missing something in our current implementation, so if you find a problem and are able to demonstrate it, we'd love to get things patched up!
